Set up code coverage

yuriidanyliak opened the pull request:
https://github.com/colszowka/simplecov
- [x] set up
- [x] run coverage report

yuriidanyliak commented:
@sergeyradkevich, created pull request. Please review.

sergeyradkevich commented:
@yuriidanyliak, this pull request should be based on master, but I see there are changes from #11. Please rebase so there are only changes related to test coverage.

sergeyradkevich commented:
@yuriidanyliak, How can I see the report (deployed)?

yuriidanyliak commented:
@sergeyradkevich, the problem is that code coverage was based on the tests written with factory_girl using. Should I leave for a pull request only deployed code coverage statistics?

sergeyradkevich commented:
@yuriidanyliak, yes, please rebase so there are only changes that are related to coverage. As I mentioned, this pull request should be based on master.

yuriidanyliak commented:
@sergeyradkevich, updated pull request; deployed coverage: [Heroku](https://mighty-tor-3045.herokuapp.com)

sergeyradkevich commented:
@yuriidanyliak, Credentials please :)

yuriidanyliak commented:
@sergeyradkevich, username: `reviewer`, password: `secret`.

yuriidanyliak commented:
Deployed version with the new way of momentary displaying of the new ToDoItem: [Heroku](https://immense-peak-9095.herokuapp.com/).

sergeyradkevich commented on app/models/to_do_list.rb:
```diff
@@ -4,6 +4,6 @@ class ToDoList < ActiveRecord::Base
   validates :name, presence: true
 
-  scope :including_items_ordered_by_id, -> { includes(:to_do_items).order(:id) }
+  scope :including_items_ordered_by_id, -> { includes(:to_do_items).order(:id).order('to_do_items.id') }
```
@yuriidanyliak, Do not see a test for the scope.

sergeyradkevich commented on test/models/to_do_list_test.rb:
```diff
@@ -24,6 +24,16 @@ class ToDoListTest < ActiveSupport::TestCase
+
+    assert_nil ToDoItem.find_by_to_do_list_id(list.id)
+
+    ToDoItem.delete_all
```
@yuriidanyliak, What is a need to delete all items here?

sergeyradkevich commented on test/models/to_do_list_test.rb:
```diff
@@ -24,6 +24,16 @@ class ToDoListTest < ActiveSupport::TestCase
+  test 'when deleting the list, its item should be also deleted' do
+    list = create(:to_do_list, id: 35)
+    create_list(:to_do_item, 10, to_do_list: list)
+    list.destroy
```
@yuriidanyliak, please ping me to clarify the process of creation of the list and its item.

sergeyradkevich commented:
@yuriidanyliak, Have you done things partially? I mean I do not see tests for the controllers to make 100% coverage.

sergeyradkevich commented:
@yuriidanyliak, Not sure whether https://mighty-tor-3045.herokuapp.com/ has been updated since the last push.

yuriidanyliak commented:
Deployed new version of the app: [Heroku](https://immense-peak-9095.herokuapp.com/).

yuriidanyliak commented:
Published code coverage statistics: [Heroku](https://mighty-tor-3045.herokuapp.com).
User name and password are the same.
Please clear the browser cache or open this page in the browser private mode if you have already opened this page in the current browser session.

yuriidanyliak commented:
@sergeyradkevich, updated, please review.

yuriidanyliak commented:
@sergeyradkevich, I'd like to mention that I nave eventually no idea how to test update action failing, when we updating the ToDoItem simply with the json which consists with the one and only boolean field. I've tried to pass the nil as this field, nil as the ToDoList and ToDoItem id, but the only result is to get the RoutingError not the error which returns when the updating is failing.

yuriidanyliak commented:
@sergeyradkevich,
1. Updated, please review.
2. [Deployed](http://immense-peak-9095.herokuapp.com/).
3. Updated [code coverage](https://mighty-tor-3045.herokuapp.com). (Login/password are still the same).

sergeyradkevich commented:
Thanks, @yuriidanyliak. Will take a look tomorrow as the first thing in the morning.

sergeyradkevich commented on test/factories.rb:
```diff
@@ -27,8 +30,10 @@
-      done nil
+    factory :done_prop_is_not_specified do
+      # done property now can not be nil because we need the correct update action errors handling
+      # if the done value is not specified the default value will be 'false'
```
@yuriidanyliak, no need of this comment. In real application it becomes irrelevant very soon. Please remove given comment.

sergeyradkevich commented on test/models/to_do_item_test.rb:
```diff
@@ -36,9 +40,10 @@ class ToDoItemTest < ActiveSupport::TestCase
   test 'new ToDoItem should create when the done property is not specified' do
-    item = create(:done_prop_is_nil)
+    # changed this test according to app behaviour changing
+    item = create(:done_prop_is_not_specified)
```
@yuriidanyliak, please avoid such comments. It doesn't bring any value into the code. Please read up the style guide about the commnets: https://github.com/bbatsov/ruby-style-guide#comments

sergeyradkevich commented on test/models/to_do_item_test.rb:
```diff
@@ -46,8 +51,13 @@ class ToDoItemTest < ActiveSupport::TestCase
+  test 'ToDoItems should be ordered by id using the scope' do
+
+    create_list(:to_do_item, 5, :random_id)
+    item_ids = ToDoItem.ids
```
@yuriidanyliak, Please remove unnecessary empty line.

sergeyradkevich commented on test/models/to_do_list_test.rb:
```diff
@@ -24,8 +29,28 @@ class ToDoListTest < ActiveSupport::TestCase
+    assert_equal list_ids.sort, sorted_list_ids
+
+    create_list(:to_do_item, 5, :random_id, to_do_list: ToDoList.first)
+    item_ids = ToDoItem.ids
```
@yuriidanyliak, this test needs to be splitted into two ones. The test should be small and cover single behavior. Please extract the second part of the test into separate one.

sergeyradkevich commented on test/models/to_do_list_test.rb:
```diff
@@ -24,8 +29,28 @@ class ToDoListTest < ActiveSupport::TestCase
+
+    assert_equal item_ids.sort, sorted_item_ids
+
+
```
And please remove empty line.

sergeyradkevich commented:
@yuriidanyliak, reviewed.

yuriidanyliak commented:
@sergeyradkevich, Updated.

sergeyradkevich commented on db/migrate/20150603090734_disallow_nil_values_for_done_prop_in_to_do_items.rb:
```diff
@@ -0,0 +1,6 @@
+class DisallowNilValuesForDonePropInToDoItems < ActiveRecord::Migration
+  def change
+    ToDoItem.connection.execute('UPDATE to_do_items SET done = false WHERE done IS NULL')
+    change_column :to_do_items, :done, :boolean, null: false
```
@yuriidanyliak, `execute` method is available in scope of migration, no need to call through `ToDoItem.connection`.

FYI: Another way to do a massive updates is to use http://apidock.com/rails/v4.0.2/ActiveRecord/Relation/update_all instead of raw sql.

sergeyradkevich commented:
@yuriidanyliak, reviewed.

yuriidanyliak commented:
@sergeyradkevich, updated.
